[AMDGPU] Added hot-block-rematerialize pass

adam-yang opened the pull request:
This is a reduced PR from https://github.com/llvm/llvm-project/pull/126331.

HotBlockRematerialize (HBR) is an optimization pass for reducing register pressure and increasing occupancy by sinking and cloning instructions.

In the original change, there are a few different code paths to target different ways of reducing register pressure and option toggles to control aggressiveness. This change only includes the default and aggressive SGPR code path.

By default, if the max SGPR count is near the wave limit (and is at risk of spilling into VGPR) or occupancy is limited by SGPR, HBR tries to reduce SGPR count by sinking or cloning single scalar instructions to their users.

adam-yang committed:
Adding remat piece by piece

adam-yang committed:
First build

adam-yang committed:
Tests

adam-yang committed:
Added test for the phi crash in pressure tracker

adam-yang committed:
clang format

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUOccupancyAndLatencyHelper.cpp:
```diff
@@ -0,0 +1,166 @@
+    MachineBasicBlock &MBB = MFI;
+    MachineBasicBlock::iterator Next;
+    AMDGPULatencyTracker LatencyTracker(ST);
+    for (auto &MI : MBB) {
+      LatencyTracker.scan(MI);
+    }
```
```suggestion
    for (auto &MI : MBB)
      LatencyTracker.scan(MI);
```

adam-yang replied:
Did a style pass.

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUOccupancyAndLatencyHelper.cpp:
```diff
@@ -0,0 +1,166 @@
+      LatencyTracker.scan(MI);
+    }
+    unsigned LoopDepth = 0;
+    if (MLI) {
+      LoopDepth = MLI->getLoopDepth(&MBB);
+    }
```
```suggestion
    if (MLI)
      LoopDepth = MLI->getLoopDepth(&MBB);
```

shiltian commented on llvm/lib/Target/AMDGPU/AMDGPUOccupancyAndLatencyHelper.cpp:
```diff
@@ -0,0 +1,166 @@
+    case AluStatus::Scalar: {
+      Score.Alu += Latency;
+      // Ignore mix alu.
+      if (PrevStatus != Status) {
+        PrevStatus = AluStatus::Nothing;
+      } else {
+        Score.MixAlu += Latency;
+      }
```
```suggestion
      if (PrevStatus != Status)
        PrevStatus = AluStatus::Nothing;
      else
        Score.MixAlu += Latency;
```

shiltian reviewed (commented):
Just a quick drive-by with some surface-level style comments. Can you go through your entire PR and make sure everything conforms to LLVM coding style?

rampitec commented:
Is there any data supporting this improves performance? How much does it hurt and where?

rampitec commented:
It will also benefit some description of the idea.

adam-yang committed:
LLVM Style

adam-yang commented:
> Just a quick drive-by with some surface-level style comments. Can you go through your entire PR and make sure everything conforms to LLVM coding style?

Did a style pass.

adam-yang commented:
> Is there any data supporting this improves performance? How much does it hurt and where?

For our hw's, the default, non-aggressive SGPR code path (which is included in this change) is enabled by default. If we turn it off, across our retail shader test set, occupancy goes down across the board. Among the shaders that showed any change in performance, occupancy lowered by ~1 on average.

For real timing measurement on PIX captures, the total frame time of all our performance suite goes down by 1% with the pass turned off.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUHotBlockRematerialize.cpp:
```diff
@@ -0,0 +1,1511 @@
+
+static cl::opt<bool>
+    EnableAggressive("amdgpu-remat-enable-hot-block-remat-aggressive");
+static cl::opt<unsigned> TargetOccupancy("amdgpu-remat-target-occupancy");
```
Avoid more cl::opts. There is already a target occupancy from the IR

adam-yang replied:
Could you point me to it? I couldn't find the right keywords.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUHotBlockRematerialize.cpp:
```diff
@@ -0,0 +1,1511 @@
+  const TargetRegisterClass *RC = MRI.getRegClass(Reg);
+  // When the unique def has subReg, just create newReg for the subReg part.
+  bool IsSubRegDef = false;
+  if (DefOp.getSubReg() != 0) {
+    RC = SIRI->getSubRegisterClass(RC, DefOp.getSubReg());
```
Recreating getOpRegClass?

adam-yang replied:
Yes basically. Replaced.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUHotBlockRematerialize.cpp:
```diff
@@ -0,0 +1,1511 @@
+    RC = SIRI->getSubRegisterClass(RC, DefOp.getSubReg());
+    IsSubRegDef = true;
+  }
+  const DebugLoc DL = DefMI->getDebugLoc();
```
```suggestion
  const DebugLoc &DL = DefMI->getDebugLoc();
```

adam-yang replied:
Fixed.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUHotBlockRematerialize.cpp:
```diff
@@ -0,0 +1,1511 @@
+//===----------------------------------------------------------------------===//
+//
+/// \file
+/// \brief AMDGPU hot block Rematerialize
```
Can we do this in the IR? The complexity is about 100x higher to do this on MachineIR while dealing with LiveIntervals

adam-yang replied:
Agreed, it would make certain things much easier. It's been tried, but measuring liveness was an issue. It was also difficult to get an accurate picture of the register pressure, partly because the distinction between uniform vector operations and scalar operations was unclear.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUHotBlockRematerialize.cpp:
```diff
@@ -0,0 +1,1511 @@
+  return false;
+}
+
+// SGPR has alignment requirment, cannot get accurate reg number.
```
The alignment requirement should not matter for spill purposes, we're going to be copying 32-bits at a time. Is this about allocation granularity? Shouldn't hardcode that either

adam-yang replied:
It's referring to allocation granularity, but in reality it's just a heuristic bias number. Changed its name to reflect this fact.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUHotBlockRematerialize.cpp:
```diff
@@ -0,0 +1,1511 @@
+  for (MachineInstr &UseMI : MRI.use_nodbg_instructions(Reg)) {
+    BBSet.insert(UseMI.getParent());
+  }
+  if (BBSet.size() == 0)
```
```suggestion
  if (BBSet.empty())
```

arsenm commented on llvm/test/CodeGen/AMDGPU/remat/simple_sgpr.mir:
```diff
@@ -0,0 +1,452 @@
+    ; undef %1.sub2:sgpr_128 = COPY $sgpr6
+    ; undef %1.sub3:sgpr_128 = COPY $sgpr7
+
+    %3000:sgpr_128 = S_LOAD_DWORDX4_IMM %0:sgpr_64, 0, 0
```
Compact register numbers

arsenm commented on llvm/test/CodeGen/AMDGPU/remat/simple_sgpr.mir:
```diff
@@ -0,0 +1,452 @@
+    ret void
+  }
+  attributes #1 = { "target-cpu"="gfx1010" }
+  !llvm.ident = !{!0}
```
Drop unnecessary annotations

arsenm reviewed (commented):
Only started skimming, but this is a huge amount of code and disproportionately large compared to the new tests. Why is this done while maintaining LiveIntervals in codegen? We know which blocks are hot in the IR, can't we just do trivial block cloning there?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUOccupancyAndLatencyHelper.cpp:
```diff
@@ -0,0 +1,162 @@
+  }
+}
+
+SchedScore collectLatency(MachineFunction &MF, const llvm::GCNSubtarget &ST,
```
This is reinventing MachineTraceMetrics?

adam-yang replied:
Partly. It has GPU (or even AMDGPU) specific metrics that MachineTraceMetrics won't give us.

adam-yang committed:
Avoid duplicate shadow variable names

adam-yang committed:
Big cleanup to clarify the flow of data, the purpose of functions, etc

adam-yang committed:
Deleted the isLocal* functions from Utils

adam-yang committed:
Fixed SCC, and updated and simplified tests

adam-yang committed:
call empty instead of size==0

adam-yang commented on llvm/lib/Target/AMDGPU/AMDGPUMIRUtils.cpp:
```diff
@@ -0,0 +1,657 @@
+} // namespace llvm
+
+namespace {
+bool isLocalSegment(const LiveRange::Segment *Seg, SlotIndexes *Indexes,
```
Yes, removed this version.

adam-yang commented on llvm/lib/Target/AMDGPU/AMDGPUMIRUtils.cpp:
```diff
@@ -0,0 +1,657 @@
+
+namespace llvm {
+
+bool isSccLiveAt(llvm::MachineBasicBlock *MBB,
```
Rewrote this whole SCC situation to use LiveIntervals instead, and added tests to verify the functionality.

adam-yang commented on llvm/lib/Target/AMDGPU/AMDGPUMIRUtils.cpp:
```diff
@@ -0,0 +1,657 @@
+      return It->getIterator();
+  }
+
+  // If no safe location can be found in the block we can save and restore
+  // SCC around MI. There is no way to directly read or Write SCC so we use
+  // s_cselect to read the current value of SCC and s_cmp to Write the saved
+  // value back to SCC.
```
Done.

adam-yang commented on llvm/lib/Target/AMDGPU/AMDGPUMIRUtils.cpp:
```diff
@@ -0,0 +1,657 @@
+}
+
+static std::vector<unsigned>
+getMinimalSpanningSubRegIdxSetForLaneMask(const TargetRegisterInfo *TRI,
```
This is removed, along with the remove lanes functionality.

adam-yang commented on llvm/lib/Target/AMDGPU/AMDGPUMIRUtils.cpp:
```diff
@@ -0,0 +1,657 @@
+
+bool isSub0Sub1SingleDef(unsigned Reg, const MachineRegisterInfo &MRI) {
+  // Support multi def for pattern of pointer:
+  // undef_ %808.sub0:sgpr_64 = COPY killed %795:sgpr_32
+  // %808.sub1:sgpr_64 = S_MOV_B32 0
```
> This is removed, along with the remove lanes functionality.

adam-yang commented on llvm/lib/Target/AMDGPU/AMDGPUHotBlockRematerialize.cpp:
```diff
@@ -0,0 +1,1511 @@
+  RematNode(unsigned R, MachineInstr *MI, unsigned S)
+      : Reg(R), DefMI(MI), InsertBlock(nullptr), InsertPointMI(nullptr),
+        Kind(RematKind::Candidate), Size(S) {}
+  unsigned Reg;
```
Done.

adam-yang commented on llvm/lib/Target/AMDGPU/AMDGPUMIRUtils.cpp:
```diff
@@ -0,0 +1,657 @@
+  return true;
+}
+
+bool isLocalLiveRange(const LiveRange *Range, SlotIndexes *Indexes,
```
Yes, removed this version.

adam-yang commented:
Latest changes:
- Deleted the whole code path for removing unused lanes. There's still some unresolved issues in there, and we could add it back in a separate change.
- Rewrote much of the SCC lifetime handling to use LiveIntervals instead of manual computation. Also added tests to exercise this functionality.
- Additional comments and clean-up to make purpose of helper functions and dataflow clearer.
